p2p/src/net/mock/transport/tcp/mod.rs

```diff
@@ -40,58 +48,130 @@ use crate::{
     P2pError, Result,
 };
 
+use self::adapter::StreamAdapter;
+
+// How much time is allowed to spend setting up (optionally) encrypted stream.
+const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(10);
+
 #[derive(Debug)]
-pub struct TcpMockTransport {}
+pub struct TcpMockTransport<E: StreamAdapter>(PhantomData<E>);
 
 #[async_trait]
-impl MockTransport for TcpMockTransport {
+impl<E: StreamAdapter + 'static> MockTransport for TcpMockTransport<E> {
     type Address = SocketAddr;
     type BannableAddress = IpAddr;
-    type Listener = TcpListener;
-    type Stream = TcpMockStream;
+    type Listener = TcpMockListener<E>;
+    type Stream = TcpMockStream<E>;
+    type StreamKey = E::StreamKey;
 
-    async fn bind(address: Self::Address) -> Result<Self::Listener> {
-        TcpListener::bind(address).await.map_err(Into::into)
+    async fn bind(stream_key: &Self::StreamKey, address: Self::Address) -> Result<Self::Listener> {
+        TcpMockListener::start(stream_key, address).await
     }
 
-    async fn connect(address: Self::Address) -> Result<Self::Stream> {
-        let stream = TcpStream::connect(address).await?;
-        Ok(TcpMockStream::new(stream))
+    async fn connect(stream_key: &Self::StreamKey, address: Self::Address) -> Result<Self::Stream> {
+        let base = TcpStream::connect(address).await?;
+        let stream = TcpMockStream::new(stream_key, base, Role::Outbound).await?;
+        Ok(stream)
+    }
+}
+
+pub struct TcpMockListener<E: StreamAdapter> {
+    receiver: UnboundedReceiver<(TcpMockStream<E>, SocketAddr)>,
+    local_address: SocketAddr,
+    join_handle: tokio::task::JoinHandle<()>,
+    _phantom: PhantomData<E>,
+}
+
+impl<E: StreamAdapter + 'static> TcpMockListener<E> {
+    async fn start(stream_key: &E::StreamKey, address: SocketAddr) -> Result<Self> {
```

Comment:
Why pass the key by reference then clone it?

Rule of thumb: If you pass something by reference then clone it, then it makes sense to just pass it by value.

Reply:
Updated, thanks.